luavm/src/lib/lua_values/mod.rs

```diff
@@ -155,6 +178,69 @@ impl LuaVal {
     }
 }
 
+impl PartialEq for LuaVal {
+    fn eq(&self, other: &LuaVal) -> bool {
+        if self.is_number() && other.is_number() {
+            return self.to_float().unwrap() == other.to_float().unwrap();
+        } else if self.is_string() && other.is_string() {
+            return self.to_string().unwrap() == other.to_string().unwrap();
+        } else if self.kind() == other.kind() {
+            if self.kind() == LuaValKind::NIL {
+                return true;
+            } else if self.kind() == LuaValKind::TABLE {
+                return unsafe { (*table_ptr(self.val)).same_ptr(&*table_ptr(other.val)) };
+            }
+        }
+        false
+    }
+}
+
+impl Eq for LuaVal {}
+
+impl Hash for LuaVal {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self.kind() {
+            LuaValKind::INT | LuaValKind::FLOAT | LuaValKind::NIL | LuaValKind::TABLE => {
+                // because val is tagged, this means that an int's hash will never collide
+                // with a table's hash, so this is safe to use
+                self.val.hash(state)
+            }
+            LuaValKind::BOXED => {
+                let val = unsafe { &*self.as_boxed() };
+                if val.is_string() {
+                    val.to_string().unwrap().hash(state)
+                } else if val.is_float() {
+                    let f = val.to_float().unwrap();
+                    // NaN or Infs cannot be hashed, and Lua doesn't handle them either
+                    if f.is_nan() || f.is_infinite() {
+                        panic!("Tried to hash NaN or Inf")
+                    }
+                    // We have to be very careful here when hashing
+                    // Consider the following example: we have two LuaVals
+                    // LuaVal { val: 0xABC3 } // This is a TABLE
+                    // LuaVal { val: 0xABC0 } // This is a BOXED LuaVal, which points to
+                    // a LuaInt { v: 0xABC3 }
+                    // Let's assume that the hash of the first LuaVal is `X`.
+                    // If we calculate the hash of the second LuaVal, we have to follow
+                    // the pointer, because we are interested in the actual value being
+                    // pointed to. If we follow the Box, and hash the contents, we will
+                    // get `X` again. This is not what we wanted, because we would like
+                    // to get different results if the types are different.
+                    // This is why in the case of Boxed Ints or Floats we add some extra
+                    // data. This ensures that collisions won't happen when two vals are
+                    // not equal.
```

Comment:
I'm going to admit that I'm baffled by this. First, it's not a problem if distinct hashes collide: indeed, we *expect* it to happen, given that an effectively infinite amount of data will be hashed down to a single machine word. Second, doing `to_string()` is going to be super slow (a heap based allocation), when we expect hashing to be super fast.

Reply:
You are right. Hopefully I fixed this in the new commit!

Reply:
Actually, I still use `to_string` for the `LuaString` case. I will need a bit more time to think about it.

Reply:
The last commit should improve the performance quite a bit. I removed all unnecessary heap allocations (hopefully).